src/transformers/generation/candidate_generator.py

```diff
@@ -149,9 +151,14 @@ def __init__(
         self.generation_config.output_scores = True
 
         # avoid unnecessary warnings that min_length is larger than max_new_tokens
+        # remove the `MinLengthLogitsProcessor` if exists (NOTE: no need to check for `MinNewTokensLogitsProcessor`)
         self.main_model_min_length = self.generation_config.min_length
         self.generation_config.min_length = 0
         self.generation_config.min_new_tokens = None
+        for processor in self.logits_processor:
+            if type(processor) == MinLengthLogitsProcessor:
+                self.main_model_min_length = getattr(processor, "min_length")
+                self.logits_processor.remove(processor)
```

Comment:
hmmm - this seems pretty hacky. Why are we removing the logits processor like this?

Reply:
This was causing errors because we are overriding the "min_length" and passing it as a kwarg into `generate` every call. If we do not remove, there will be two min lengths, one as a processor and one as a kwarg.

I am not sure right now how to tackle this another way, I'll give it a thought. I did not think user can rewrite `MinLengthLogitsProcessor` their own way

Reply:
No, you are right, that should be possible. I'll check

Reply:
After a bit of digging, I think that removing the  `MinLengthLogitsProcessor` is the only solution because we are allowing users to pass in the same thing either as a kwarg or as a "transformers-existing LogitsProcessor". Regardless of way they pass it, the same generation is done, but using "kwargs" is a more straightforward way. Passing LogitsProcessorList can be enabled only for custom processors, raising warnings/errors if transformer-existing processor is passed. 

@gante wdyt? this will be a breaking change but might ease maintaining in the long run.

Reply:
I'd rather we just raised an exception if we detect both the kwarg and the logit processor are set. Otherwise we can get unexpected behaviour where the logitprocessor's min_length is ignored. 

This raises a more general question about the responsibilities of logit processors v.s. kwargs - it seems there might be future cases when there's overlap. 

For what it's worth - I'm not sure how much sense there is in having a min length logits processor? I'd expect the processors to modify the scores, but not necessarily to control the generation logic.

Reply:
> For what it's worth - I'm not sure how much sense there is in having a min length logits processor? I'd expect the processors to modify the scores, but not necessarily to control the generation logic.

Agreed here, but I am not sure it's a stopping criteria either. 

> This raises a more general question about the responsibilities of logit processors v.s. kwargs - it seems there might be future cases when there's overlap.

Hmm, overlap in a sense that both (kwargs and processors) perform same change in logits? I believe that using logits processors is for advanced users and custom cases, so not sure about possible overlaps

Reply:
> Agreed here, but I am not sure it's a stopping criteria either.

I think it can be interpreted as a stopping criterion i.e. if it hasn't be satisfied the generation shouldn't stop. 

> Hmm, overlap in a sense that both (kwargs and processors) perform same change in logits? 

Overlap in the sense that they can both have parameters which control the same thing. I'm not sure about changing logits here - is this needed for specifying a min length? 

> I believe that using logits processors is for advanced users and custom cases, so not sure about possible overlaps

It's more of a question of whether our API allows it - we should be wary of Hyrum's law!

Reply:
@amyeroberts @zucchini-nlp Answering a question above

> For what it's worth - I'm not sure how much sense there is in having a min length logits processor? I'd expect the processors to modify the scores, but not necessarily to control the generation logic.

The min length processor does modify the scores, it sets the probability of EOS tokens to 0. Generation always stops when EOS is selected, thus we have to modify the probabilities before selecting the token so as to enforce a minimum length. If we didn't modify the probabilities, then the generation loop would have to become more complex (if eos was selected and we had a minimum length, we would need to go back and pick a different token). 

Remember also that there can be multiple stopping criteria in addition to the EOS token (maximum length is reached, a wall clock time limit was reached, ...), and that generation stops when one of them is met -- preventing generation from stopping until `min_length` is reached would break this behavior 🤗

___________________________________________________

Regarding the code changes:

`MinLengthLogitsProcessor` can only be present in `self.logits_processor` if a user manually passes a `MinLengthLogitsProcessor` instance to `generate` (as opposed to the documented `min_length` or `min_new_tokens` arguments). By being present as a custom processor, the candidate generation process generates more candidates than it should, hurting performance. Note that the overlap is okay (two minimum length processors is equivalent to only have the largest minimum length), but the custom `MinLengthLogitsProcessor` is not.

Having any form of logits processing in assisted generation is not a hard requirement, but rather a performance-enhancing operation -- it forces the assistant model to have the same logic processing as the main model. Sadly, Whisper relies on custom logits processors with assisted generation to secure the speedups announced in our Distilwhisper paper, so we can't simply remove support for them to get rid of these (and future related) issues at the cost of a minor throughput hit.

As Amy wrote, exceptions > handling ourselves. I would raise an exception if either `MinLengthLogitsProcessor` or `MinNewTokensLengthLogitsProcessor` are set as custom logits processors.

Reply:
> As Amy wrote, exceptions > handling ourselves. I would raise an exception if either MinLengthLogitsProcessor or MinNewTokensLengthLogitsProcessor are set as custom logits processors.

Yes, raising error was the second option I had, it just seemed a bit un-intuitive from user point to be able to pass LogitsProcessor and then getting an error. Okay, I will raise exception and change the doctests then :)